File: notebooksjava/_05RemoveDuplicateNodes.java
import java.util.HashSet;
import java.util.Set;

public class _05RemoveDuplicateNodes {

    static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
    }

    // HashSet で訪問済みの値を管理し、重複ノードをスキップする
    public ListNode removeDuplicateNodes(ListNode head) {
        if (head == null) return null;

        Set<Integer> visited = new HashSet<>();
        visited.add(head.val);
        ListNode curr = head;

        while (curr != null && curr.next != null) {
            if (visited.contains(curr.next.val)) {
                curr.next = curr.next.next;
            } else {
                visited.add(curr.next.val);
                curr = curr.next;
            }
        }
        return head;
    }

    // ── 補助関数 ──────────────────────────────────────────────

    static ListNode createList(int... vals) {
        if (vals.length == 0) return null;
        ListNode head = new ListNode(vals[0]);
        ListNode curr = head;
        for (int i = 1; i < vals.length; i++) {
            curr.next = new ListNode(vals[i]);
            curr = curr.next;
        }
        return head;
    }

    static int[] toArray(ListNode head) {
        int size = 0;
        ListNode curr = head;
        while (curr != null) { size++; curr = curr.next; }
        int[] arr = new int[size];
        curr = head;
        for (int i = 0; i < size; i++) { arr[i] = curr.val; curr = curr.next; }
        return arr;
    }

    static boolean arrayEquals(int[] a, int[] b) {
        if (a.length != b.length) return false;
        for (int i = 0; i < a.length; i++) if (a[i] != b[i]) return false;
        return true;
    }

    public static void main(String[] args) {
        _05RemoveDuplicateNodes sol = new _05RemoveDuplicateNodes();

        // null head
        assert sol.removeDuplicateNodes(null) == null;

        // no duplicates
        assert arrayEquals(toArray(sol.removeDuplicateNodes(createList(1, 2, 3))), new int[]{1, 2, 3});

        // duplicates at end
        assert arrayEquals(toArray(sol.removeDuplicateNodes(createList(1, 2, 3, 3, 2, 1))), new int[]{1, 2, 3});

        // all same
        assert arrayEquals(toArray(sol.removeDuplicateNodes(createList(5, 5, 5))), new int[]{5});

        // single node
        assert arrayEquals(toArray(sol.removeDuplicateNodes(createList(7))), new int[]{7});

        System.out.println("All test cases passed ✅");
    }
}

File: notebooksc++/05-remove-duplicate-nodes.cpp
#include <iostream>
#include <unordered_set>
#include <vector>
#include <cassert>

using namespace std;

// Definition for singly-linked list
struct ListNode {
    int val;
    ListNode* next;
    ListNode(int val = 0, ListNode* next = nullptr) : val(val), next(next) {}
};

// Remove duplicate nodes from linked list
ListNode* removeDuplicateNodes(ListNode* head) {
    if (head == nullptr) {
        return head;
    }
    
    unordered_set<int> visited;
    visited.insert(head->val);
    ListNode* curr = head;
    
    while (curr != nullptr && curr->next != nullptr) {
        if (visited.count(curr->next->val) > 0) {
            curr->next = curr->next->next;
        } else {
            visited.insert(curr->next->val);
            curr = curr->next;
        }
    }
    
    return head;
}

// Helper function to create linked list from vector
ListNode* createLinkedList(const vector<int>& values) {
    if (values.empty()) {
        return nullptr;
    }
    
    ListNode* head = new ListNode(values[0]);
    ListNode* curr = head;
    
    for (size_t i = 1; i < values.size(); i++) {
        curr->next = new ListNode(values[i]);
        curr = curr->next;
    }
    
    return head;
}

// Helper function to convert linked list to vector
vector<int> linkedListToVector(ListNode* head) {
    vector<int> result;
    ListNode* curr = head;
    
    while (curr != nullptr) {
        result.push_back(curr->val);
        curr = curr->next;
    }
    
    return result;
}

// Helper function to clean up memory
void deleteLinkedList(ListNode* head) {
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

int main() {
    // Test 1: list with duplicates 1 -> 2 -> 3 -> 3 -> 2 -> 1
    ListNode* head1 = createLinkedList({1, 2, 3, 3, 2, 1});
    ListNode* result1 = removeDuplicateNodes(head1);
    vector<int> vec1 = linkedListToVector(result1);
    vector<int> expected1 = {1, 2, 3};
    assert(vec1 == expected1);
    deleteLinkedList(result1);
    
    // Test 2: list with no duplicates 1 -> 2 -> 3
    ListNode* head2 = createLinkedList({1, 2, 3});
    ListNode* result2 = removeDuplicateNodes(head2);
    vector<int> vec2 = linkedListToVector(result2);
    vector<int> expected2 = {1, 2, 3};
    assert(vec2 == expected2);
    deleteLinkedList(result2);
    
    // Test 3: empty list
    ListNode* head3 = removeDuplicateNodes(nullptr);
    assert(head3 == nullptr);
    
    // Test 4: single element
    ListNode* head4 = createLinkedList({5});
    ListNode* result4 = removeDuplicateNodes(head4);
    vector<int> vec4 = linkedListToVector(result4);
    vector<int> expected4 = {5};
    assert(vec4 == expected4);
    deleteLinkedList(result4);
    
    cout << "All test cases passed ✅" << endl;
    return 0;
}
